### ISSUE TITLE
Provide support for schemas scoped by a query

### ISSUE THREAD

**Issue by zachdaniel** [elixir-ecto/ecto#1797]
One of the best things about Ecto is the separation of concerns that it offers with regards to your application's data, and that data's representation in the application layer.  One thing that we've noticed that has drastically cleaned up design throughout our application is creating new, first class objects to represent different 'kinds' of entities.  A simple example would be having a `users` table, with a `kind` field, that could be set to either "student" or "teacher".  Our ideal way to present this to the rest of the application is as three separate modules: User, Student, and Teacher.  This allows us to limit what fields show up on each struct, as well as have different changesets/validations for each.  This helps us get around a very classic problem about scrubbing certain kinds of fields for certain kinds of users, and prevent data leakage.  When requesting an instance of `User`, the current user is unauthorized unless they are logged in as that user.  However, when requesting an instance of `Student` the current user is authorized if they are a teacher of that student.  There are many other similar cases that we see in our application.

tl;dr - having schemas scoped by a query can help clean up the design of a system, and simplify a whole class of data representation problems. We could absolutely manage our own views, but I can say from experience that using migrations to manage/update views/DML can be come a huge pain point.

----

While the above would be nice, we may actually end up using views, due to the limitations of the above approach.  What would fit very nicely into what we're looking for would be integrating the creation/management of views into Ecto migrations. With Postgres' features of inserting directly into views, one could treat this view exactly like a table.  What I suspect will be the real difficulty in the implementation will be the restrictions that postgres places on views in order for them to be "automatically updatable".  The restrictions follow:

* The view must have exactly one entry in its FROM list, which must be a table or another updatable view.
* The view definition must not contain WITH, DISTINCT, GROUP BY, HAVING, LIMIT, or OFFSET clauses at the top level.
* The view definition must not contain set operations (UNION, INTERSECT or EXCEPT) at the top level.
* The view's select list must not contain any aggregates, window functions or set-returning functions.

I'm happy to discuss other potential design decisions, but I've already seen a massive clarity/generecism boost in our code base using this pattern, (even though we are faking it a bit).

Thank you for your time 😄

**Comment by zachdaniel** [comment]
Although I'm fairly new to elixir+ecto, its possible I may end up taking a stab at this.

**Comment by josevalim** [comment]
I don't believe we would be able to support this, except for every specific cases which would definitely be more limited than PG views.

There are different ways you could solve this problem in Ecto. For example, you could define the schemas as:

    schema "unset_users" do
      ...
    end

And now, if you try to do `Repo.all(User)`, it will fail. This gives you data protection. You can now export all, teacher, student, etc accordingly:

    def students do
      from u in {"users", __MODULE__}, where: u.type == "student"
    end

and so on.

The other idea is to make it easier to build views. Views expect a query and we have a great feature for building queries, what if we allow views to be created with Ecto's query syntax?

**Comment by zachdaniel** [comment]
I can definitely see the value in allowing views to be created from Ecto queries, but that still doesn't really solve the problem of always having to store the full up/down of an entire view in every migration that alters it.  A simple example:

```
#create_students.ex
defmodule Migrations.CreateStudent do
  user Ecto.Migration

  def up do
     create view(:students) do
       from user in MyApp.User,
         where: user.type == "student",
         select: [:username, :age]
     end
  end

  def down do
    drop_view(:students)
  end
end

# Want to add a field?
defmodule Migrations.AddFieldToStudent do
  def up do
    create view(:students) do
       from user in MyApp.User,
         where: user.type == "student",
         select: [:username, :age, :new_field]
     end
  end

  def down do
     create_or_replace view(:students) do
       from user in MyApp.User,
         where: user.type == "student",
         select: [:username, :age]
     end
  end
end
```

This sort of thing can get tedious and annoying quickly, especially when the number of views/iterations on views rises.  It might be nice to have a simplified form of view management, that takes things like equality restrictions on columns, or perhaps a series of `where` statements and a list of columns?  Would be great to be able to treat them more like tables in that fashion.  Things could start looking like this:

```
defmodule Migrations.AddFieldToStudent do
  def change(user) do
    add :field, user.some_new_field
    add :filter, user.age < 21
  end
end
```

I can already see a lot of holes in that kind of design, but I wonder if we might be able to get close?  This pattern could really help people keep their data model thin and optimized and get them leveraging some of the more useful features that postgres offers (views as an API for your data).

**Comment by josevalim** [comment]
> I can already see a lot of holes in that kind of design

Yeah, unless Postgres supports something like that, I am not sure we can support it with sane semantics or without doing tons of work. I think supporting queries will at least make things a bit more digestable.

**Comment by zachdaniel** [comment]
It would be possible to get something a little bit in between if we added a few features to composable queries.  For instance, if you could select in pieces, deselect, tag certain filters, and use those tags to remove them later, then a series of migrations of a view could be represented as ongoing alterations of a query.

like
```
#migration 1

create view(:students, from: :users) do
  from row in source,
    tag: :student_type_filter,
    where: row.type == "student",
    select: [:username, :email]
end

#migration 2

alter view(:students) do
  from row in source,
    remove_tag: :student_type_filter,
    deselect: [:email],
    tag: :age_filter,
    where: row.age < 21
end
```

Obviously a lot of contrived examples, and I'm pretty sure it might be one of those things where the juice isn't worth the squeeze.  Just wanted to illustrate a way that makes it potentially possible (as well as making it slightly easier to compose super complex queries with lots of conditional joins.

**Comment by zachdaniel** [comment]
In fact, I don't know if it is a good idea to generate views w/ the ecto query builder.  Lets say you have a one year old migration that runs a query over your `MyApp.User` model, you run into huge issues as soon as you want to remove or rename that model, in that your history is no longer consistent.  We made this same mistake at my last company.  Its good practice to never reference application code from migrations, because migrations must be timeless, but application code is only current.

**Comment by josevalim** [comment]
> Obviously a lot of contrived examples, and I'm pretty sure it might be one of those things where the juice isn't worth the squeeze.

The problem is that we would need to store those views and queries in the database, because we don't quite now which query is currently in the database (and we don't load all previous migrations as well so we can't know from migrations).

> In fact, I don't know if it is a good idea to generate views w/ the ecto query builder.

Oh, if you are using Ecto queries to build views, you must not access your app code. Just write the ecto query without calling pieces of your app.

**Comment by josevalim** [comment]
Ok, so I would like to propose a possible solution to your problem that relies on composable queries. The idea is to have a single schema but build the filters using query composition:

```elixir
defmodule User do
  use Ecto.Schema

  schema "unset_users" do
    field :type, :string
    ...
  end

  def all(query \\ __MODULE__) do
    {"users", query}
  end

  def get(query \\ __MODULE__, type) do
    from u in {"users", query}, where: [type: ^type]
  end

  def build(type) do
    %User{type: type} |> Ecto.put_meta(source: "users")
  end
end
```

The idea is to build a schema that you cannot really query until choose a type. Once the type is chosen, you set the proper schema. This approach will work on most cases since Ecto allows you to pass queries for composition. For example, if you add "has_many :users, User" to another schema, you can further filter it when preloading:

    from s in School, preload: [users: ^User.get(:students)]

To get all users regardless of type:

    from s in School, preload: [users: ^User.all]

The only downside of this approach is that you cannot load students, teachers and other roles side by side. But then I would argue a view would be preferable otherwise you would be doing multiple queries over the same table, every time looking for a different type.

Everything you need for the approach above to work is on Ecto master.

**Comment by josevalim** [comment]
Ok, besides the changes above, I have also created another issue for views: #1805. We can close this one now (discussion can continue in the separate issues). :)

**Comment by MarthinL** [comment]
Hey,

I found this seemingly abandoned idea while looking for a declarative way apply fixed filter to a table source over which multiple schemas are defined in manner similar to polymorphism but at schema level rather than association level where it can easily be done with where: clauses. I'd like to avoid views and I don't want an entire query to be the source of the schema, but I want to nominate a discriminator field and value that would be added to the SQL whenever the schema is used and potentially even forcing the field to the designated value on inserts and updates.

I noticed in the code that the schema currently strictly allows a string value for source, as in:
````
defmodule MyApp.Post
use Ecto.Schema

schema "bases" do
    field :part, Ecto.Enum, values: [post: 3, comment: 4]
    field :text, :string, source; :value
end
````

Meaning the programmer has to find a way to ensure that everywhere Post is used to include where: [part: post] or p.part == ^:post depending if assoc or query syntax is being used.

Naively it seems that it would be fairly easy and safe to upgrade Ecto to allow a 2-tuple in the place of the source string, such as:
````
defmodule MyApp.Post
use Ecto.Schema

schema {"bases", [part: :post]} do
    field :part, Ecto.Enum, values: [post: 3, comment: 4]
    field :text, :string, source; :value
end
````
This filter would probaby mean an additional field in the metadata the code that composes the SQL for the schema would use to initialise the list of filter conditions with.

This way, anywhere where such a schema is used its guaranteed to address only records from the underlying table that matches the filter condition including assoc definitions.

It might be possible, advisable or even requried to omit the discriminator field completely out of the schema and/or add it automatically as a view-only field.

````
defmodule MyApp.Post
use Ecto.Schema

schema {"bases", [part: 3]} do
    field :text, :string, source; :value
end
````

What say you, could this be something worth doing?

**Comment by josevalim** [comment]
Ecto has a really neat feature, in that it allows you to replace the source of any query. So if you want the behaviour you are asking, I'd simply use database views. For example, you can have

```
defmodule Post do
  schema "bases_part_3" do
    ...
  end
end
```

where `bases_part_3` is a database view on top of the actual table called "bases". Now, if you want to drop the view, you can simply do `from p in {"bases", Post}`. I show this technique in practice here: https://dashbit.co/blog/soft-deletes-with-ecto

**Comment by MarthinL** [comment]
Using views does not yield the same result because:
a) the programmer still needs to ensure that the discriminator (part) is correctly set at all times. Automatically updatable views apply but depending on the specific settings for the view it would either allow changing the row to values that wouldn't pass the filter anymore or simply refuse to make the change, but it won't automatically set the field because the database does not know that it's that simple.

b) The views would have to be handcrafted and would be side-effects of migrations which are carefully controlled to only run once. By handling it at the schema level it is exactly as dynamic as the schema definitions themselves and does not require database changes.

c) Managing views and even using Ecto queries syntax for view definition was something you considered for a while back in 2016 but then abandoned apparently because of just too many problems such as the ones I've listed above.

d) One of the coolest things in Phoenix, especially now with LiveView is how allows programmers to solve the entire problem from the same language domain. I was a database-centric person first and wrote entire systems in Oracle PLSQL in my time and still I consider expressing what I need from the database in Elixir with Ecto.Query syntax a major benefit to myself and those whoever will need to look after my systems after me. I'm not against leaving the environment when it makes sufficient sense, like coding a database extension in C (which I've done) or adding specialised graphics to the web-page in JavaScript (which I've done too) but then those should ideally be neatly fenced concerns which seamlessly tie back into the Phoenix domain for then bulk of the business logic to use.

I know you from reading many of your posts that you actually subscribe to the same values and ideas as I laid out above. Which is why I don't understand the change of heart and suggesting views. But OK then, I can and have solved my problem less elegantly but still effectively without views already. I honestly thought such a clean and complete solution to a problem you have written about before would be welcomed rather than dismissed. You live and learn. Won't try it again.

**Comment by josevalim** [comment]
The problems with views was generating them automatically from Ecto queries, not the views themselves.

>  I honestly thought such a clean and complete solution to a problem you have written about before would be welcomed rather than dismissed. You live and learn. Won't try it again.

Your idea was not dismissed. It was read, considered, but not an approach we want to go with. Hearing "no" is part of open source. Plus the issue is 8 year old: myself and anyone else is entitled to change their mind within that period (and even shorter ones). :) Have a good one.

**Comment by zachdaniel** [comment]
I'd like to be clear that I'm about to express how you can do this with Ash, but Ash is not an *alternative* to Ecto, but rather a higher level abstraction that may or may not be what you want. I also want to make it clear that Ash wasn't built out of any kind of "disappointment" with what can be accomplished with Ecto. Ecto powers all of Ash's sql querying logic, meaning that it is, in fact, a *love letter* to Ecto ❤️

I also don't support your harsh response to @josevalim's expression of how one ought to solve this problem with Ecto, and would encourage you not to assume bad faith on others' part, especially not @josevalim 😁. 

With all that said, you can do this with an `Ash.Resource`, using the `base_filter` option. 

```elixir
defmodule MyApp.Post
  use Ash.Resource,
    data_layer: AshPostgres.DataLayer

  actions do
    defaults [:read]
  end

  resource do
    base_filter part: :post
  end

  postgres do
    table "bases"
    repo YourRepo
  end

  attribute do
    attribute :part, YourEnumType
    attribute :text, :string, source: :value
  end
end
```

Again, the Ecto project has my full support, my only motivation in posting this is to express that perhaps the kind of thing you are looking for sits at a higher level of abstraction than Ecto, and could be found in Ash. At least form my perspective :)

@josevalim if you think its inappropriate for me to post this kind of thing here, I'll remove it or feel free to remove it yourself. Not interested in astroturfing Ecto's issues 🙇

**Comment by MarthinL** [comment]
@zachdaniel thanks for pointing out the apparent harshness of my response. 

@josevalim I'm sorry if you also thought I overstepped some boundary. It was without malice or suspicion of bad faith. 

Contrary to how my response seems to come across its directness and honesty is actually evident of the extremely high regard I hold for @josevalim and his creations. I was an early Erlang adopter and evangelist and am still inspired to tackle issues head on and stick to reasoned arguments by the example set by Joe Armstrong who always impressed with the strength of his vision and willingness to accommodate new ideas. 

I've a lot to learn about Ash before I can pass any comment on it, and I'm not in a good place with my project to be retooling to that extent even if the few things I've seen about it seem to indicate that Ash might have served my purposes well if not better. I've long preferred model-based concepts.

What I can say however is that while I get why you might say the solution I am seeking sits at a higher level of abstraction than Ecto offers I do not agree with that assessment. Filtering records at the schema level is exactly at the same level of abstraction as filtering associations which are very elegantly done using where: in the association declaration. To me its omission feels like an understandable oversight rather than an add-on feature on a stable platform. When I read the discussion about such features from 2016 and 2018 its usefulness wasn't disputed but nobody seemed to have an active use-case for it at the time so it died a quiet death. That's very understandable and I fully support the general principle that one should never implement anything for which you don't have an active and motivated user to test it against. 


What I tried saying is that I have the real use case now and saw a way to do a safe, compliant, complete and tested implementation of a version of the functionality that was contemplated in 2016. Ecto from 3.0 was marked stable, not stagnant, right. 😏

But it's really not a big deal if the general sentiment is to leave Ecto be and focus on new stuff like Ash. If that's how everyone feels in due course I too will make my way over to Rome to make like the Romans.
